File: Algorithms/C++/33-Search_in_Rotated_Sorted_Array.cpp
// Runtime: 4 ms, faster than 82.06% of C++ online submissions for Search in Rotated Sorted Array.
// Memory Usage: 8.8 MB, less than 69.75% of C++ online submissions for Search in Rotated Sorted Array.
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (nums[mid] == target)
                return mid;
            if (nums[mid] >= nums[l]) { // left continue
                if (target >= nums[l] && target < nums[mid])
                    r = mid - 1;
                else
                    l = mid + 1;
            } else { // right continue
                if (target > nums[mid] && target <= nums[r])
                    l = mid + 1;
                else
                    r = mid - 1;
            }
        }
        return -1;
    }
};

File: Algorithms/Python3.x/33-Search_in_Rotated_Sorted_Array.py
class Solution:
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        start, end = 0, len(nums) - 1
        while start <= end:
            mid = (start + end) // 2
            if nums[mid] == target: return mid  # 找到
            if nums[mid] > nums[end]:
                if nums[start] <= target < nums[mid]:
                    end = mid - 1
                else:
                    start = mid + 1
            else:  # nums[mid] < nums[end]
                if nums[mid] < target <= nums[end]:
                    start = mid + 1
                else:
                    end = mid - 1
        return -1


'''
分为4种情况：
    中间的大于后面的：
        目标值在左半部分
        目标值在右半部分
    中间的小于后面的：（数组中不存在重复项，所以没有等于）
        目标值在右半部分
        目标值在左半部分
'''
